_source/_posts/2023-01-23-keycloak-kubernetes-prod.md

```diff
@@ -0,0 +1,644 @@
+                  * Handler that will be invoked when this action is resuming after an external redirect. If your         
+                  * onExecutePostLogin function does not perform a redirect, this function can be safely ignored.         
+                  *                                                                                                       
+                  * @param {Event} event - Details about the user and the context in which they are logging in.           
+                  * @param {PostLoginAPI} api - Interface whose methods can be used to change the behavior of the login.  
+                  */                                                                                                      
+                 // exports.onContinuePostLogin = async (event, api) => {                                                 
+                 // };                                                                                                    
+
+```
+
+You can list the available actions with the following command:
+
+```shell
+auth0 actions list
+```
+The output will show the deployment status of each action:
+
+```text
+=== dev-avup2laz.us.auth0.com actions
+
+  ID                                    NAME      TYPE        STATUS  DEPLOYED  
+  ee0ea308-0e50-4be3-893b-74f9ccbb3703  jhipster  post-login  built   ✗         
+```
+
+Note the `DEPLOYED` status is `x`. Go ahead and deploy it using the action ID:
+
+```shell
+auth0 actions deploy ee0ea308-0e50-4be3-893b-74f9ccbb3703
+```
+
+Once the action is deployed, you must attach it to the login flow. Sign in to Auth0, and on the left menu choose **Actions**. Then in the Flows screen, choose **Login**. On the right bottom section, choose **Custom**. Drag the **jhipster** action to the flow diagram. Then click **Apply**. The flow should look like the following:
+
+{% img blog/keycloak-kubernetes-prod/login-flow.png alt:"Custom Auth0 Login Action" width:"500" %}{: .center-image }
+
+Now that all is set on the identity provider side, let's configure Keycloak as an identity broker delegating authentication to Auth0.
+
+### Add the Auth0 Identity Provider
+
+Navigate to **http://keycloak.\<namespace\>.\<public-ip\>.nip.io**, and sign in with admin/admin. On the welcome page, choose **Administration Console**. On the left menu, for the top options,  choose the **jhipster** realm. Then, at the bottom of the menu, choose **Identity providers**. In the User-defined section, choose **OpenID Connect v1.0**. Fill in the provider configuration as follows:
+
+- RedirectURI: **(pre-filled)**
+- Alias: **auth0**
+- User discovery endpoint: **On**
+- Discovery endpoint: **https://\<auth0-domain\>/.well-known/openid-configuration**
+- Client authentication: **Client secret sent as post**
+- Client ID: **auth0-client-id**
+- Client Secret: **auth0-client-secret**
```

Comment:
The default is "algorithm not specified", which seems kinda scary.

Reply:
I added a note about **client assertion signature algorithm**. I believe it is not required for the sleected client authentication method **client secret sent as post**. Also it seems Auth0 free plan does not support client authentication with JWTs. Please let me know if my assumptions are ok.

Reply:
I asked a colleague about the Auth0 free plan and client authentication, and he said "All of our authentication flows use JWT regardless of plan".

Reply:
In Auth0 dashboard, I saw something different in the application details, credentials tab. The "Private Key JWT" option is disabled and it has a label "Enterprise".